folding-schemes/src/folding/protogalaxy/mod.rs

```diff
@@ -76,3 +232,809 @@ pub enum ProtoGalaxyError {
+            cf_u_i1_x = self.cf_U_i.hash_cyclefold(&sponge, self.pp_hash);
+
+            augmented_F_circuit = AugmentedFCircuit::empty(
+                &self.poseidon_config,
+                self.F.clone(),
+                self.U_i.betas.len(),
+                d,
+                k,
+            );
+            augmented_F_circuit.pp_hash = self.pp_hash;
+            augmented_F_circuit.z_0.clone_from(&self.z_0);
+            augmented_F_circuit.z_i.clone_from(&self.z_i);
+            augmented_F_circuit
+                .external_inputs
+                .clone_from(&external_inputs);
+
+            // There is no need to update `self.U_i` etc. as they are unchanged.
+        } else {
+            // Primary part:
+            // Compute `U_{i+1}` by folding `u_i` into `U_i`.
+            let (U_i1, W_i1, F_coeffs, K_coeffs, L_evals, phi_stars) = Folding::prove(
+                &mut transcript_prover,
+                &self.r1cs,
+                &self.U_i,
+                &self.W_i,
+                &[self.u_i.clone()],
+                &[self.w_i.clone()],
+            )?;
+
+            // CycleFold part:
+            // get the vector used as public inputs 'x' in the CycleFold circuit
+            // cyclefold circuit for cmW
+            let mut r0_bits = L_evals[0].into_bigint().to_bits_le();
+            let mut r1_bits = L_evals[1].into_bigint().to_bits_le();
+            r0_bits.resize(C1::ScalarField::MODULUS_BIT_SIZE as usize, false);
+            r1_bits.resize(C1::ScalarField::MODULUS_BIT_SIZE as usize, false);
+
+            let cfW_u_i_x = [
+                r0_bits
+                    .chunks(C1::BaseField::MODULUS_BIT_SIZE as usize - 1)
+                    .map(BigInteger::from_bits_le)
+                    .map(C1::BaseField::from_bigint)
+                    .collect::<Option<Vec<_>>>()
+                    .unwrap(),
+                get_cm_coordinates(&C1::zero()),
+                get_cm_coordinates(&self.U_i.phi),
+                get_cm_coordinates(&phi_stars[0]),
+            ]
+            .concat();
+            // cyclefold circuit for cmE
+            let cfE_u_i_x = [
+                r1_bits
+                    .chunks(C1::BaseField::MODULUS_BIT_SIZE as usize - 1)
+                    .map(BigInteger::from_bits_le)
+                    .map(C1::BaseField::from_bigint)
+                    .collect::<Option<Vec<_>>>()
+                    .unwrap(),
+                get_cm_coordinates(&phi_stars[0]),
+                get_cm_coordinates(&self.u_i.phi),
+                get_cm_coordinates(&U_i1.phi),
+            ]
+            .concat();
+
+            let cfW_circuit = ProtoGalaxyCycleFoldCircuit::<C1, GC1> {
+                _gc: PhantomData,
+                r_bits: Some(vec![r0_bits.clone()]),
+                points: Some(vec![C1::zero(), self.U_i.phi]),
+                x: Some(cfW_u_i_x.clone()),
+            };
+            let cfE_circuit = ProtoGalaxyCycleFoldCircuit::<C1, GC1> {
```

Comment:
I'm not sure I'm following this part, from where the 2 CycleFold circuits of W and E come from? in ProtoGalaxy we should only need 1 single CycleFold circuit to check the fold of the $\phi$ points, no?
In a similar way as in HyperNova we only use a single CycleFold circuit to fold the multiple $C$ points: 
https://github.com/privacy-scaling-explorations/sonobe/tree/7097c001fc876578be2229d3590d506858bc0069/folding-schemes/src/folding/hypernova/mod.rs#L692

Reply:
Sorry for the confusion caused! `cfW` and `cfE` are copied from Nova, but I forgot to rename them. And yeah using only 1 CycleFold circuit is definitely better, given that we already support multiple points in CycleFold circuits.

Reply:
In https://github.com/privacy-scaling-explorations/sonobe/pull/123/commits/3191a92919a4633387840c611e6058c1bf0bb4a9, I renamed `cfW_circuit` to `cf1_circuit`, and `cfE_circuit` to `cf2_circuit`.

However, it is a bit tricky to use a single CF circuit to handle the computation of `U_{i + 1}.phi = U_i.phi * L_evals[0] + u_i.phi * L_evals[1]`. Although the CF circuit now supports adding multiple points, it still lacks support for multiple randomness...

I have some ideas for improving CycleFold, in terms of the in-circuit logic and the creation of public inputs `x`, which should address this and @CPerezz's comments on the conversion from `r_bits` to `r_fp` in `x`.

Maybe we can have a separate PR for this, what do you think?

Reply:
Oh right! I was influenced by HyperNova, but in ProtoGalaxy the values that multiply each instance are not just powers of a random value, so we can not use the tricks used for CycleFold in HyperNova with multiple instances :S

Then, the current implementation in this PR supports folding 2-to-1 instances but not k-to-1, right? Independent of this, I agree with leaving it as it is now in this PR, and the CycleFold improvements that you mention can be done in future iterations.

Reply:
(regarding multiple-instances folding (k-to-1), I just see this comment and if mentioning it: https://github.com/winderica/sonobe/blob/d779889b8f05069f4c0a0cbf2aa0134c28c01aed/folding-schemes/src/folding/protogalaxy/mod.rs#L580 , cool)

Reply:
Yeah only 2-to-1 folding is supported by the current implementation, but I left some space for extending it to k-to-1 folding.

In fact, the parameters of the methods of `AugmentationGadget` can already have arbitrary length, and in the future we only need to connect `other_instances` in `prove_step` with the variables in `AugmentedFCircuit`.
https://github.com/privacy-scaling-explorations/sonobe/blob/d779889b8f05069f4c0a0cbf2aa0134c28c01aed/folding-schemes/src/folding/protogalaxy/circuits.rs#L131-L217